src/server/terrain/simple_structures/SbbGen.zig

```diff
@@ -0,0 +1,95 @@
+const std = @import("std");
+
+const main = @import("main");
+const terrain = main.server.terrain;
+const Vec3i = main.vec.Vec3i;
+const GenerationMode = terrain.biomes.SimpleStructureModel.GenerationMode;
+const CaveMapView = terrain.CaveMap.CaveMapView;
+const CaveBiomeMapView = terrain.CaveBiomeMap.CaveBiomeMapView;
+const sbb = terrain.structure_building_blocks;
+const Blueprint = main.blueprint.Blueprint;
+const ZonElement = main.ZonElement;
+const Neighbor = main.chunk.Neighbor;
+const ServerChunk = main.chunk.ServerChunk;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+const parseBlock = main.blocks.parseBlock;
+
+pub const id = "cubyz:sbb";
+pub const generationMode = .floor;
+
+const SubstitutionMap = std.AutoHashMapUnmanaged(u16, u16);
+
+const SbbGen = @This();
+
+structure: []const u8,
+structureRef: *const sbb.StructureBuildingBlock,
+placeMode: Blueprint.PasteMode,
+
+pub fn getHash(self: SbbGen) u64 {
+	return std.hash.Wyhash.hash(@intFromEnum(self.placeMode), self.structure);
+}
+
+pub fn loadModel(arenaAllocator: NeverFailingAllocator, parameters: ZonElement) *SbbGen {
+	const structureId = parameters.get(?[]const u8, "structure", null) orelse unreachable;
+	const structureRef = sbb.getByStringId(structureId) orelse {
+		std.log.err("Could not find structure building block with id '{s}'", .{structureId});
+		unreachable;
+	};
+	const self = arenaAllocator.create(SbbGen);
+	self.* = .{
+		.structure = structureId,
+		.structureRef = structureRef,
+		.placeMode = std.meta.stringToEnum(Blueprint.PasteMode, parameters.get([]const u8, "placeMode", "degradable")) orelse Blueprint.PasteMode.degradable,
+	};
+	return self;
+}
+
+pub fn generate(self: *SbbGen, _: GenerationMode, x: i32, y: i32, z: i32, chunk: *ServerChunk, _: CaveMapView, _: CaveBiomeMapView, seed: *u64, _: bool) void {
+	placeSbb(self, self.structureRef, Vec3i{x, y, z}, Neighbor.dirUp, chunk, seed);
+}
+
+fn placeSbb(self: *SbbGen, structure: *const sbb.StructureBuildingBlock, placementPosition: Vec3i, placementDirection: Neighbor, chunk: *ServerChunk, seed: *u64) void {
+	const origin = structure.blueprints[0].originBlock;
+	const rotationCount = alignDirections(origin.direction(), placementDirection) catch |err| {
+		std.log.err("Could not align directions {s} and {s} error: {s}", .{@tagName(origin.direction()), @tagName(placementDirection), @errorName(err)});
```

Comment:
Not a blocker for this PR, but in the future we definitely need a better error message for this, including the names of the structures.